src/workerd/io/worker-fs.c++

```diff
@@ -0,0 +1,1319 @@
+  return kj::rc<FileImpl>(kj::mv(owned));
+}
+
+kj::Rc<File> File::newReadable(kj::ArrayPtr<const kj::byte> data) {
+  return kj::rc<FileImpl>(data);
+}
+
+kj::Own<VirtualFileSystem> newVirtualFileSystem(
+    kj::Own<FsMap> fsMap, kj::Rc<Directory>&& root, kj::Own<VirtualFileSystem::Observer> observer) {
+  return kj::heap<VirtualFileSystemImpl>(kj::mv(fsMap), kj::mv(root), kj::mv(observer));
+}
+
+kj::Own<VirtualFileSystem> newWorkerFileSystem(kj::Own<FsMap> fsMap,
+    kj::Rc<Directory> bundleDirectory,
+    kj::Own<VirtualFileSystem::Observer> observer) {
+  // Our root directory is a read-only directory
+  Directory::Builder builder;
+  builder.addPath(fsMap->getBundlePath(), kj::mv(bundleDirectory));
+  builder.addPath(fsMap->getTempPath(), getTmpDirectoryImpl());
+  builder.addPath(fsMap->getDevPath(), getDevDirectory());
+  return newVirtualFileSystem(kj::mv(fsMap), builder.finish(), kj::mv(observer));
+}
+
+kj::Rc<Directory> getTmpDirectoryImpl() {
+  return kj::rc<TmpDirectory>();
+}
+
+bool TmpDirStoreScope::hasCurrent() {
+  return tmpDirStorageScope != nullptr;
+}
+
+TmpDirStoreScope& TmpDirStoreScope::current() {
+  KJ_ASSERT(hasCurrent(), "no current TmpDirStoreScope");
+  return *static_cast<TmpDirStoreScope*>(tmpDirStorageScope);
+}
+
+TmpDirStoreScope::TmpDirStoreScope(kj::Maybe<kj::Badge<TmpDirStoreScope>> guard)
+    : dir(Directory::newWritable()) {
+  if (guard == kj::none) {
+    kj::requireOnStack(this, "must be created on the stack");
+    onStack = true;
+    KJ_ASSERT(!hasCurrent(), "TmpDirStoreScope already exists on this thread");
+    tmpDirStorageScope = this;
+  }
+}
+
+TmpDirStoreScope::~TmpDirStoreScope() noexcept(false) {
+  if (onStack) {
+    KJ_ASSERT(tmpDirStorageScope == this, "this TmpDirStoreScope not on the stack");
+    tmpDirStorageScope = nullptr;
+  }
+}
+
+kj::Own<TmpDirStoreScope> TmpDirStoreScope::create() {
+  // Creating the instance with the badge will ensure that
+  // it is not set as current in the stack.
+  return kj::heap<TmpDirStoreScope>(kj::Badge<TmpDirStoreScope>());
+}
+
+Stat SymbolicLink::stat(jsg::Lock& js) {
+  return Stat{
+    .type = FsType::SYMLINK,
+    .size = 0,
+    .lastModified = kj::UNIX_EPOCH,
+    .writable = false,
+  };
+}
+
+jsg::Url SymbolicLink::getTargetUrl() const {
+  auto path = getTargetPath().toString(false);
+  return KJ_ASSERT_NONNULL(jsg::Url::tryParse(path, "file:///"_kj));
+}
+
+kj::Maybe<kj::OneOf<kj::Rc<File>, kj::Rc<Directory>>> SymbolicLink::resolve(jsg::Lock& js) {
+  KJ_IF_SOME(ret, root->tryOpen(js, getTargetPath())) {
+    KJ_SWITCH_ONEOF(ret) {
+      KJ_CASE_ONEOF(file, kj::Rc<File>) {
+        return kj::Maybe<kj::OneOf<kj::Rc<File>, kj::Rc<Directory>>>(kj::mv(file));
+      }
+      KJ_CASE_ONEOF(dir, kj::Rc<Directory>) {
+        return kj::Maybe<kj::OneOf<kj::Rc<File>, kj::Rc<Directory>>>(kj::mv(dir));
+      }
+      KJ_CASE_ONEOF(link, kj::Rc<SymbolicLink>) {
+        // A symbolic link won't resolve to another symbolic link.
```

Comment:
Why not?

Reply:
The `resolve(...)` method here follows all symbolic links in the path, so when it encounters a symlink as the path is being processed, it will attempt to resolve it into it's target or return `kj::none`. If you want the symlink itself, then use `tryOpen(...)` on the directory

Reply:
Sweet, makes sense, would love to see this answer inlined into the comment next time you edit this code.